src/Microsoft.AspNetCore.Razor.OmniSharpPlugin/DefaultProjectChangePublisher.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.IO;
+using Microsoft.AspNetCore.Razor.OmniSharpPlugin.StrongNamed.Serialization;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Microsoft.AspNetCore.Razor.OmniSharpPlugin
+{
+    [Shared]
+    [Export(typeof(ProjectChangePublisher))]
+    [Export(typeof(OmniSharpProjectSnapshotManagerChangeTrigger))]
+    internal class DefaultProjectChangePublisher : ProjectChangePublisher, OmniSharpProjectSnapshotManagerChangeTrigger
+    {
+        private readonly ILogger<DefaultProjectChangePublisher> _logger;
+        private readonly JsonSerializer _serializer;
+        private readonly Dictionary<string, string> _publishFilePathMappings;
+        private readonly object _publishLock;
+        private OmniSharpProjectSnapshotManagerBase _projectManager;
+
+        [ImportingConstructor]
+        public DefaultProjectChangePublisher(ILoggerFactory loggerFactory)
+        {
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger<DefaultProjectChangePublisher>();
+
+            _serializer = new JsonSerializer()
+            {
+                Formatting = Formatting.Indented,
+            };
+            _serializer.Converters.RegisterOmniSharpRazorConverters();
+            _publishFilePathMappings = new Dictionary<string, string>(FilePathComparer.Instance);
+            _publishLock = new object();
+        }
+
+        public void Initialize(OmniSharpProjectSnapshotManagerBase projectManager)
+        {
+            if (projectManager == null)
+            {
+                throw new ArgumentNullException(nameof(projectManager));
+            }
+
+            _projectManager = projectManager;
+            _projectManager.Changed += ProjectManager_Changed;
+        }
+
+        public override void SetPublishFilePath(string projectFilePath, string publishFilePath)
+        {
+            lock (_publishLock)
+            {
+                _publishFilePathMappings[projectFilePath] = publishFilePath;
+            }
+        }
+
+        // Virtual for testing
+        protected virtual void DeleteFile(string publishFilePath)
+        {
+            var info = new FileInfo(publishFilePath);
+            if (info.Exists)
+            {
+                info.Delete();
```

Comment:
The parent locks it. I can rename `DeleteFileUnsafe`

Reply:
That would be fine for private methods. I prefer moving the lock here because this method is protected.

Reply:
Sure that's fine too. Only protected for testing though.

Reply:
Actually can't do that. The lock is primarily for accessing the publish file mappings not for the actual deletion. I will add a try catch around the delete logic in case there's a race between the exists and the delete though